File: ios-swiftUI/Tday/Core/Data/List/ListShareRepository.swift
import Foundation

/// Whether a shared-list id refers to a scheduled list or a floater list.
enum ShareListKind {
    case scheduled
    case floater

    /// API path segment ("list" or "floaterList").
    var pathBase: String {
        switch self {
        case .scheduled: return "list"
        case .floater: return "floaterList"
        }
    }
}

/// Membership management for shared lists. Deliberately ONLINE-ONLY:
/// membership changes are rare, identity-sensitive operations with ugly
/// offline-replay failure modes (silently re-adding a removed member,
/// demote/remove races), so they never enqueue pending mutations —
/// connectivity errors surface directly.
@MainActor
final class ListShareRepository {
    private let api: TdayAPIService

    init(api: TdayAPIService) {
        self.api = api
    }

    func fetchMembers(kind: ShareListKind, listID: String) async throws -> ListMembersResponse {
        try await api.getListMembers(base: kind.pathBase, listID: listID)
    }

    func addMember(kind: ShareListKind, listID: String, username: String, role: String) async throws -> ListMemberDTO? {
        try await api.addListMember(
            base: kind.pathBase,
            listID: listID,
            payload: AddMemberRequest(username: username, role: role)
        ).member
    }

    func updateMemberRole(kind: ShareListKind, listID: String, userId: String, role: String) async throws {
        _ = try await api.updateListMemberRole(
            base: kind.pathBase,
            listID: listID,
            payload: UpdateMemberRoleRequest(userId: userId, role: role)
        )
    }

    func removeMember(kind: ShareListKind, listID: String, userId: String) async throws {
        _ = try await api.removeListMember(
            base: kind.pathBase,
            listID: listID,
            payload: RemoveMemberRequest(userId: userId)
        )
    }

    func leaveList(kind: ShareListKind, listID: String) async throws {
        _ = try await api.leaveList(base: kind.pathBase, listID: listID)
    }

    func searchUsers(query: String) async throws -> [UserSearchResultDTO] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else { return [] }
        return try await api.searchUsers(query: trimmed).users
    }
}

File: android-compose/app/src/main/java/com/ohmz/tday/compose/core/data/list/ListShareRepository.kt
package com.ohmz.tday.compose.core.data.list

import com.ohmz.tday.compose.core.data.requireApiBody
import com.ohmz.tday.compose.core.model.AddMemberRequest
import com.ohmz.tday.compose.core.model.ListMemberDto
import com.ohmz.tday.compose.core.model.ListMembersResponse
import com.ohmz.tday.compose.core.model.RemoveMemberRequest
import com.ohmz.tday.compose.core.model.UpdateMemberRoleRequest
import com.ohmz.tday.compose.core.model.UserSearchResultDto
import com.ohmz.tday.compose.core.network.TdayApiService
import javax.inject.Inject
import javax.inject.Singleton

/** Whether a shared-list id refers to a scheduled list or a floater list. */
enum class ShareListKind { SCHEDULED, FLOATER }

/**
 * Membership management for shared lists. Deliberately ONLINE-ONLY: membership
 * changes are rare, identity-sensitive operations with ugly offline-replay
 * failure modes (silently re-adding a removed member, demote/remove races), so
 * they never enqueue pending mutations — connectivity errors surface directly.
 */
@Singleton
class ListShareRepository @Inject constructor(
    private val api: TdayApiService,
) {
    suspend fun fetchMembers(kind: ShareListKind, listId: String): ListMembersResponse =
        requireApiBody(
            when (kind) {
                ShareListKind.SCHEDULED -> api.getListMembers(listId)
                ShareListKind.FLOATER -> api.getFloaterListMembers(listId)
            },
            "Could not load members",
        )

    suspend fun addMember(
        kind: ShareListKind,
        listId: String,
        username: String,
        role: String
    ): ListMemberDto? =
        requireApiBody(
            when (kind) {
                ShareListKind.SCHEDULED -> api.addListMember(
                    listId,
                    AddMemberRequest(username = username, role = role)
                )

                ShareListKind.FLOATER -> api.addFloaterListMember(
                    listId,
                    AddMemberRequest(username = username, role = role)
                )
            },
            "Could not add member",
        ).member

    suspend fun updateMemberRole(
        kind: ShareListKind,
        listId: String,
        userId: String,
        role: String
    ) {
        requireApiBody(
            when (kind) {
                ShareListKind.SCHEDULED -> api.updateListMemberRole(
                    listId,
                    UpdateMemberRoleRequest(userId = userId, role = role)
                )

                ShareListKind.FLOATER -> api.updateFloaterListMemberRole(
                    listId,
                    UpdateMemberRoleRequest(userId = userId, role = role)
                )
            },
            "Could not update member role",
        )
    }

    suspend fun removeMember(kind: ShareListKind, listId: String, userId: String) {
        requireApiBody(
            when (kind) {
                ShareListKind.SCHEDULED -> api.removeListMember(
                    listId,
                    RemoveMemberRequest(userId = userId)
                )

                ShareListKind.FLOATER -> api.removeFloaterListMember(
                    listId,
                    RemoveMemberRequest(userId = userId)
                )
            },
            "Could not remove member",
        )
    }

    suspend fun leaveList(kind: ShareListKind, listId: String) {
        requireApiBody(
            when (kind) {
                ShareListKind.SCHEDULED -> api.leaveList(listId)
                ShareListKind.FLOATER -> api.leaveFloaterList(listId)
            },
            "Could not leave list",
        )
    }

    suspend fun searchUsers(query: String): List<UserSearchResultDto> {
        if (query.trim().length < 2) return emptyList()
        return requireApiBody(api.searchUsers(query.trim()), "Could not search users").users
    }
}
